### PR TITLE
Add environment

### PR DESCRIPTION
**Pull request by Kuroshii**
#### Changes:
- Add an initial environment
- Update enums
- Fix tests
- Add random direction to board generation
- Add death sense

### CODE REVIEW

File: src/environment.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
+import actions
+import cell_types
+import percepts
+import board_state as state
+
+ADJACENT_BOARD_PERCEPTS = {
+    cell_types.PIT:    percepts.BREEZE,
+    cell_types.WUMPUS: percepts.STENCH
+}
+ON_SPOT_BOARD_PERCEPTS = {
```
Comment (sagersmith8) [inline src/environment.py:10]: Seems weird to split these

File: src/board_state.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+class BoardState:
+    def __init__(self, board, pos, direction):
+        self.board = board
+        self.pos = list(pos)
+        self.direction = direction
+        self.arrows = self.count_wumpuses()
+
+    def show(self):
+        for row in self.board:
```
Comment (sagersmith8) [inline src/board_state.py:31]: No docstring?

File: src/board_state.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            print ''.join(map(str, row))
+
+        print 'Position: {}'.format(self.pos)
+        print 'Direction: {}'.format(directions.NAMES[self.direction])
+        print 'Arrows: {}'.format(self.arrows)
+
+    def get_board_percepts(self):
+        current_cell = self.cell_at(self.pos)
```
Comment (sagersmith8) [inline src/board_state.py:43]: No docstring?

File: src/board_state.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        - Remove the stench from adjacent cells
+        """
+        wumpus_cell = self.cell_at(wumpus_pos)
+        wumpus_cell.cell_type = cell_types.EMPTY
+
+        for adj_cell in self.adj_cells(wumpus_pos):
+            adj_cell.remove_percept(percepts.STENCH)
+
+    def adj_cells(self, pos):
+        return (self.cell_at(move(pos, direction))
```
Comment (sagersmith8) [inline src/board_state.py:114]: No Docstrings?

File: src/board_state.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        :returns: Nothing, but changes the agent's direction
+        """
+        self.direction = (self.direction - 1) % len(directions.DIRECTIONS)
+
+    def kill_wumpus(self, wumpus_pos):
+        """
+        Update the board state for killing a wumpus a the given position.
+        - Change the cell where the wumpus was to empty
+        - Remove the stench from adjacent cells
+        """
```
Comment (sagersmith8) [inline src/board_state.py:98]: No params?

File: src/board_state.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        for adj_cell in self.adj_cells(wumpus_pos):
+            adj_cell.remove_percept(percepts.STENCH)
+
+    def adj_cells(self, pos):
+        return (self.cell_at(move(pos, direction))
+                for direction in directions.DIRECTIONS
+                if self.on_board(move(pos, direction)))
+
+    def cell_at(self, pos):
+        if not self.on_board(pos):
```
Comment (sagersmith8) [inline src/board_state.py:128]: No docstrings?

File: src/environment.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        Tells whether the agent has reached the goal state.
+        That is, tells whether the agent has picked up the gold.
+
+        :rtype: bool
+        :returns: whether the agent has picked up the gold yet
+        """
+        return self.finished
+
+    def named_percepts(self):
+        return map(
```
Comment (sagersmith8) [inline src/environment.py:76]: No docstrings?

File: src/environment.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+    ACTION_METHOD = {
+        actions.LEFT:    _do_turn_left,
+        actions.RIGHT:   _do_turn_right,
+        actions.FORWARD: _do_move_forward,
+        actions.SHOOT:   _do_shoot,
+        actions.GRAB:    _do_grab
+    }
+
+    def perform_action(self, action):
```
Comment (sagersmith8) [inline src/environment.py:232]: No docstrings?

File: src/environment.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        self._record_action(action)
+        Environment.ACTION_METHOD[action](self)
+        self._update_turn(action)
+
+    def perform_actions(self, actions):
```
Comment (sagersmith8) [inline src/environment.py:241]: No docstrings?

File: src/environment.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        for action in actions:
+            self.perform_action(action)
+
+    def turn_left(self):
```
Comment (sagersmith8) [inline src/environment.py:249]: No docstrings?

File: src/environment.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        self.perform_action(actions.LEFT)
+
+    def turn_right(self):
```
Comment (sagersmith8) [inline src/environment.py:256]: No docstrings

File: src/environment.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        self.perform_action(actions.RIGHT)
+
+    def move_forward(self):
```
Comment (sagersmith8) [inline src/environment.py:262]: No docstrings?

File: src/environment.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        self.perform_action(actions.FORWARD)
+
+    def shoot(self):
```
Comment (sagersmith8) [inline src/environment.py:275]: No docstrings?

File: src/environment.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        self.perform_action(actions.SHOOT)
+
+    def grab(self):
```
Comment (sagersmith8) [inline src/environment.py:281]: No docstrings?

File: src/environment.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+        self.perform_action(actions.GRAB)
+
+
+KILL_AGENT_CELLS = {cell_types.PIT, cell_types.WUMPUS}
+STOP_ARROW_CELLS = {cell_types.OBSTACLE, cell_types.WUMPUS}
+
+
+def deadly(cell):
```
Comment (sagersmith8) [inline src/environment.py:293]: No docstrings?

File: src/environment.py [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+    :param cell: the cell to check for deadliness
+    :type cell: Cell
+    :rtype: bool
+    :returns: True only if an agent will die if they walk into this cell
+    """
+    return cell.cell_type in KILL_AGENT_CELLS
+
+
+def stops_arrow(cell):
```
Comment (sagersmith8) [inline src/environment.py:305]: No docstrings?

File: src/environment.py [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+    Checks if the given cell will stop an arrow that flies into it.
+    :param cell: the cell to check for arrow stopping
+    :type cell: Cell
+    :rtype: bool
+    :returns: whether or not an arrow can safely pass through this cell
+    """
+    return cell.cell_type in STOP_ARROW_CELLS
+
+
+def new_game(size):
```
Comment (sagersmith8) [inline src/environment.py:316]: No docstrings?

File: src/environment.py [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+    import generate_world
+
+    world = generate_world.generate_world(size, 0.1, 0.1, 0.1)
+    return world, Environment(world)
+
+
+def see(env):
+    env.board_state.show()
+    print env.named_percepts()
```
Comment (sagersmith8) [inline src/environment.py]: You should do:

```
print "Percepts:", env.named_percepts()
```

**Review by JWhitney406 — APPROVED** [review]
Looks good!

**Pull request merged**